bftengine/src/bftengine/DbCheckpointManager.cpp

```diff
@@ -341,16 +341,16 @@ void DbCheckpointManager::sendInternalCreateDbCheckpointMsg(const SeqNum& seqNum
   req.sender = replica_id;
   req.seqNum = seqNum;
   req.noop = noop;
-  std::vector<uint8_t> data_vec;
+  std::vector<concord::Byte> data_vec;
   concord::messages::db_checkpoint_msg::serialize(data_vec, req);
-  std::string sig(SigManager::instance()->getMySigLength(), '\0');
-  SigManager::instance()->sign(reinterpret_cast<char*>(data_vec.data()), data_vec.size(), sig.data());
-  req.signature = std::vector<uint8_t>(sig.begin(), sig.end());
+  // TODO: this signature does not appear to be validated anywhere in the codebase
+  req.signature.resize(SigManager::instance()->getMySigLength());
+  SigManager::instance()->sign(seqNum, data_vec.data(), data_vec.size(), req.signature.data());
```

Comment:
As far as I understand, the method will be called at every checkpoint. 
This method is being called as part of the stable checkpoint callbacks (see https://github.com/vmware/concord-bft/blob/c769f077da98a4ef6f5ae8fa859d4618c70552d5/bftengine/src/bftengine/ReplicaImp.cpp#L5356).
Now, changing the cryptosystem also happens on a (stable ?) checkpoint, right? if so, does the order of the callbacks matter? 
Could it be an issue if we first sign the dbcheckpoint message with the old cryptosystem and only then update the cryptosystem? 
Your comment says that there is no validation of this message anywhere in the code, but if it were, will the other replicas try to validate it with the new or with the old cryptosystem (and are the meanings)?

Reply:
There is no "changing the cryptosystem".
A new cryptosystem is created when the execution handler for the key exchange is called, say in sequence s, and say the new key pair is k.
Once checkpoint (s+(150*2)) /150 is reached, the old cryptosystem which has k is removed (in CryptoManager::OnCheckpoint).
Once a stable checkpoint is reached, we use its sequence number to decide which cryptosystem to use.
If the stable checkpoint is too far behind, this message can be signed using some different new key (k' with s' > s).